### PR TITLE
Make cherami task backend open source

### PR DESCRIPTION
**Pull request by madhuravi**
This was in the private repository earlier, copying it over to open source since cherami is open source.

### CODE REVIEW

File: mocks/modules/task/cherami/Client.go [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cherami
+
+import cherami "github.com/uber/cherami-client-go/client/cherami"
+import gocherami "github.com/uber/cherami-thrift/.generated/go/cherami"
+import mock "github.com/stretchr/testify/mock"
```
Comment (glibsm) [inline mocks/modules/task/cherami/Client.go]: I almost commented that you should combine your imports.

Then I realized that this is a generated code.

But now I'm commenting anyway about the fact that I shouldn't comment.

**Comment by madhuravi** [comment]
@aravindvs @uber-go/service-framework

File: modules/task/cherami/cherami.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+	cherami_gen "github.com/uber/cherami-thrift/.generated/go/cherami"
+	"go.uber.org/zap"
+)
+
+// TODO: Most of these will be fetched from host configuration instead of being hardcoded
+const (
+	_initialized = iota
+	_running
+	_stopped
+	_hyperbahnHostsFile = "/etc/uber/hyperbahn/hosts.json"
```
Comment (alsamylkin) [inline modules/task/cherami/cherami.go]: Looks like this should be removed :)

File: modules/task/cherami/cherami.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+// Backend holds cherami data
+type Backend struct {
+	client      cherami.Client
+	publisher   cherami.Publisher
+	consumer    cherami.Consumer
+	deliveryCh  chan cherami.Delivery
+	config      clientConfig
+	logger      *zap.Logger
+	scope       tally.Scope
+	state       int
```
Comment (alsamylkin) [inline modules/task/cherami/cherami.go]: nit: use a custom type(alias) for enums? We can do better then plain old C :)

File: modules/task/cherami/cherami.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+	for i := 0; i < b.config.ConsumeWorkerCount; i++ {
+		go b.consumeAndExecute()
+	}
+	return nil
+}
+
+func (b *Backend) consumeAndExecute() {
+	defer func() {
+		if r := recover(); r != nil {
+			ulog.Logger(context.Background()).Error(
```
Comment (alsamylkin) [inline modules/task/cherami/cherami.go]: increment a taskFailure counter?(or introduce a new one)

File: modules/task/cherami/cherami.go [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+)
+
+// TODO: Most of these will be fetched from host configuration instead of being hardcoded
+const (
+	_initialized = iota
+	_running
+	_stopped
+	_hyperbahnHostsFile = "/etc/uber/hyperbahn/hosts.json"
+	_pathPrefix         = "/uberfx_async/"
+	_errChanCount       = 50
```
Comment (alsamylkin) [inline modules/task/cherami/cherami.go]: Doesn't seem to be used.

File: modules/task/cherami/cherami.go (hunk 2 above)
Comment (madhuravi) [inline modules/task/cherami/cherami.go]: excellent catch!

File: mocks/modules/task/cherami/Client.go [hunk 6]
```diff
@@ -0,0 +1,378 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (madhuravi) [inline mocks/modules/task/cherami/Client.go]: can't find a way to generate them with a custom package name yet. might send a later PR if I find a way.

File: modules/task/cherami/cherami.go [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+	_pathPrefix = "/uberfx_async/"
+)
+
+var (
+	_hyperbahnMu         sync.RWMutex
+	_hyperbahnHostsFile  string
+	_cheramiClientFunc   = cherami.NewHyperbahnClient
+	_defaultClientConfig = clientConfig{
+		ConsumerName:       "uberfx-async",
+		PrefetchCount:      10,
```
Comment (alsamylkin) [inline modules/task/cherami/cherami.go]: Do we have a configuration option for this parameter?

File: modules/task/cherami/cherami.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+		ConsumeWorkerCount: 10,
```
Comment (madhuravi) [inline modules/task/cherami/cherami.go]: PRs are not a good place for feature requests :) I think I mentioned this to you, if you can go ahead and cut a task for this, I will prioritize.

File: modules/task/cherami/cherami.go (hunk 7 above)
Comment (madhuravi) [inline modules/task/cherami/cherami.go]: Yes, everything in the clientConfig struct is configurable.

**Pull request merged**